### DISCUSSION TITLE
[useDateRangePicker] how to prevent autofocus on closeButton in dateField?

### DISCUSSION THREAD

**Discussion by ananassta** [adobe/react-spectrum#4677] [Q&A] [answered]
There is a problem i found while working with this hook (useDateRangePicker). After entering the date, the focus is automatically transferred to the clear button, located inside the input field. What can be done about it?

Referring to the hook code, I found that the lack of focus transfer to the calendar button is spelled out directly. Can we expand this function for custom buttons passed inside the field?

<img width="389" alt="Screen Shot 2023-06-19 at 20 33 53" src="https://github.com/adobe/react-spectrum/assets/55900056/7f38b316-b83b-4321-a23d-b2df02b20590">
<img width="399" alt="Screen Shot 2023-06-19 at 20 33 23" src="https://github.com/adobe/react-spectrum/assets/55900056/fedb7f97-ce38-4d49-baf5-ebb7a9f13663">

<img width="468" alt="Screen Shot 2023-06-19 at 19 36 13" src="https://github.com/adobe/react-spectrum/assets/55900056/9565bff0-8f64-4306-92f1-c3d1733d14fe">

**Comment by snowystinger** [comment]
The button that line refers to is the popover button, not the clear button. I'm not sure where that clear button is coming from, we don't have it in any of our examples that I see.

You could make the clear button non-tabbable, using tabIndex = -1. But you'll need to provide a keyboard accessible way to also clear the field. Usually 'Escape' is good for this. This is probably the best approach.

If that doesn't work, you could try some of these other solutions.

You could move the clear button to outside of the field, or, soon, rely on Form 'reset' behavior.

If you wanted the clear button to look like it was inside, but be the last tabbable element in the component, you could possibly use CSS 'order', though that may be confusing to users who are sighted as it would appear that focus is jumping around out of order.

**Reply by ananassta** [reply]
This button is really not in your example. It's a custom solution from our designers. 
Probably, the question is that in the case of using "useDatePicker" there is no such problem with focus and the focus remains on the date, while here it moves further to the button.

**Reply by snowystinger** [reply]
The reason behind the difference is that filling in the fields will automatically advance the cursor from segment to segment. This works differently in the range picker because there are two date fields, so we need to be able to advance from the first one to the second one. Currently we do that by just checking to make sure we aren't advancing to the button as you've found. Instead, we think that we could check for spinbuttons or textboxes when trying to advance the cursor instead of just !button. Would you be interested in opening a PR?

**Reply by ananassta** [reply]
Do you want me to send a pull request to the library? 
The only idea I have is to extend the hook's props with the IDs of the elements that should be ignored by the focus and then use them in createFocusManager.

**Reply by snowystinger** [reply, marked as answer]
Sending a PR is the fastest way to get a bug fixed. Otherwise you'll have to tolerate our priorities a bit more.
The idea I shared was:
> We think that we could check for spinbuttons or textboxes when trying to advance the cursor instead of just !button

So that accept function would look more like (psuedo-ish code)
```
accept: (node) => node.role === 'spinbutton' || node.role === 'textbox'
```

**Reply by tgelu** [reply]
Just want to add that there is an example in the docs which suffers from the same issue.

Here you showcase a clear button which gets focused when last year is set https://react-spectrum.adobe.com/react-aria/DateRangePicker.html#state
